### ISSUE TITLE
Need support CPU flash attention with mask

### ISSUE THREAD

**Issue by imzhuhl** [pytorch/pytorch#112380]
### 🚀 The feature, motivation and pitch

I want use flash attention on the cpu to speed up my automatic speech recognition(ASR) model, which is transformer-based.
In practice, the audio length in a batch may be different, which corresponds to different sequence lengths. The general approach is to use a mask tensor to indicate the sequence length, and add the mask to the attenton score before calculating the softmax in SDPA. Here is a simple example:
```python
def get_mask(self, lengths, dtype=torch.float32):
    # lengths: the seq's length, shape is [batch, ]
    max_seq_len = lengths.max().item()
    row_vector = torch.arange(0, max_seq_len, 1)
    matrix = torch.unsqueeze(lengths, dim=-1)
    mask = row_vector < matrix
    mask = mask.to(dtype)
    mask_4d = 1 - mask[:, None, None, :]
    mask_4d = mask_4d * torch.finfo(dtype).min
    return mask_4d, mask

lengths = [2, 3]  # batch size is 2, first seq length is 2, second is 3
mask_4d, mask = get_mask(lengths)
print(mask)
"""
[
    [1, 1, 0],
    [1, 1, 1],
]
"""

# Assume here is a sdpa block
# query, key, value are both [batch, num_heads, max_seq_len, head_dim]
q_h = query * head_dim ** (-0.5)
attention_scores = torch.matmul(q_h, key.transpose(-1, -2))
attention_scores = attention_scores + mask_4d
attention_probs = nn.functional.softmax(attention_scores, dim=-1)
context_layer = torch.matmul(attention_probs, value)
```
As the code shows, mask4d needs to be used before softmax to eliminate erroneous results.
For now, I can use `torch._scaled_dot_product_attention_math` for this, but `torch._scaled_dot_product_flash_attention` doesn't work, it doesn't provide an api for mask.

However, when I see `flash_attention_kernel` in `FlashAttentionKernel.cpp`, which is the kernel for cpu flash attention, it could support mask just simply modify the code.

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki
